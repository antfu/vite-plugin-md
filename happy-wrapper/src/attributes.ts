import { string } from 'fp-ts'
import { pipe } from 'fp-ts/lib/function'
import type { IElement, INode } from 'happy-dom'
import { createFragment } from './create'
import { HappyMishap } from './errors'
import type { Container, DocRoot, GetAttribute, HTML } from './happy-types'
import { isDocument, isElement, isFragment } from './type-guards'
import { getNodeType, solveForNodeType, toHtml } from './utils'

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export type SetAttribute<T extends string> = (value: string) => <N extends Container | HTML>(node: N) => N

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export type SetAttributeTo<_T extends string, _V extends string> = <N extends Container | HTML>(node: N) => N

export const setAttribute = <T extends string>(
  attr: T,
): SetAttribute<T> => <V extends string>(
    value: V,
  ): SetAttributeTo<T, V> => <N extends Container | HTML>(
      node: N,
    ): N => {
    const invalidNode = (n: INode) => {
      throw new HappyMishap(`You can not use the setAttribute() utility on a node of type: "${getNodeType(n)}"`, { name: `setAttribute(${attr})(${value})(INVALID)` })
    }
    const result = solveForNodeType()
      .mirror()
      .solver({
        html: h => pipe(h, createFragment, f => setAttribute(attr)(value)(f), toHtml),
        text: t => invalidNode(t),
        node: n => invalidNode(n),
        fragment: (f) => {
          f.firstElementChild.setAttribute(attr, value)
          return f
        },
        document: (d) => {
          d.body.firstElementChild.setAttribute(attr, value)
          return d
        },
        element: (e) => {
          e.setAttribute(attr, value)
          return e
        },
      })(node)

    return result
  }

export const getAttribute = <T extends string>(attr: T): GetAttribute<T> => {
  return solveForNodeType('text', 'node')
    .outputType<string>()
    .solver({
      html: h => pipe(h, createFragment, getAttribute(attr)),
      fragment: f => f.firstElementChild.getAttribute(attr),
      document: doc => doc.body.firstElementChild.getAttribute(attr),
      element: el => el.getAttribute(attr),
    })
}

const getClass = getAttribute('class')
const setClass = setAttribute('class')
/**
 * Provides the classes defined on a given container's top level
 * element as an array of strings
 */
export const getClassList = (container: Container | HTML | null): string[] => {
  if (!container)
    return []

  return solveForNodeType().outputType<string[]>().solver({
    html: h => pipe(h, createFragment, getClassList),
    document: d => getClass(d.body.firstElementChild)?.split(/\s+/) || [],
    fragment: f => getClass(f.firstElementChild)?.split(/\s+/) || [],
    element: e => getClass(e)?.split(/\s+/) || [],
    text: (n) => { throw new HappyMishap('Passed in a text node to getClassList!', { name: 'getClassList', inspect: n }) },
    node: (n) => { throw new HappyMishap('Passed in an unknown node type to getClassList!', { name: 'getClassList', inspect: n }) },
  })(container).filter(i => i)
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export type RemoveClass<R extends string | string[]> = <C extends DocRoot | IElement | HTML>(container: C) => C

/**
 * Removes a class from the top level node of a container's body.
 *
 * Note: if the class wasn't present then no change is performed
 */
export const removeClass = <R extends string | string[]>(
  remove: R,
): RemoveClass<R> => <D extends DocRoot | IElement | HTML>(doc: D): D => {
  const current = getClass(doc)?.split(/\s+/g) || []
  const toRemove: string[] = !Array.isArray(remove) ? [remove] : remove

  const resultantClassString = Array.from(
    new Set<string>(current.filter(c => !toRemove.includes(c))),
  )
    .filter(i => i)
    .join(' ')

  return setClass(resultantClassString)(doc)
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export type AddClass<A extends string[] | string[][]> = <C extends DocRoot | IElement | HTML>(container: C) => C

/**
 * Adds a class to the top level node of a document's body.
 */
export const addClass = <A extends string[] | string[][]>(
  ...add: A
): AddClass<A> => <D extends DocRoot | IElement | HTML>(doc: D): D => {
  const toAdd = (Array.isArray(add) ? add.flat() : [add]) as string[]

  const currentClasses = getClass(doc)?.split(/\s+/g) || []
  const resultantClasses = Array.from(new Set<string>([...currentClasses, ...toAdd]))

  return setClass(resultantClasses.join(' ').trim())(doc) as D
}

export type Filter = (string | RegExp)
export type FilterCallback = (removed: string[]) => void
export type FiltersWithCallback = [FilterCallback, ...Filter[]]

function hasFilterCallback(filters: Filter[] | FiltersWithCallback): filters is FiltersWithCallback {
  return typeof filters[0] === 'function'
}

/**
 * Filters classes out from a given element using _filters_, where a filter:
 *
 * - string - when a string it will compare for a direct match
 * - RegExp - will run the RegExp's `test(class)` method
 *
 * Optionally you may pass in a callback function as the first parameter in the
 * the list and this callback will be then called with all filtered properties
 * passed to it. This is useful for creating desirable side-effects like _moving_
 * the classes to some other DOM element (for instance).
 */
export const filterClasses = <A extends Filter[] | [FilterCallback, ...Filter[]]>(
  ...args: A
) => <D extends DocRoot | IElement | HTML>(
    doc: D,
  ): D => {
  const el = isDocument(doc) || isFragment(doc)
    ? doc.firstElementChild as IElement
    : isElement(doc)
      ? doc as IElement
      : null
  if (!el)
    throw new HappyMishap('An invalid container was passed into filterClasses()!', { name: 'filterClasses', inspect: doc })

  const filters = hasFilterCallback(args) ? args.slice(1) as Filter[] : args as Filter[]
  const cb = hasFilterCallback(args) ? args[0] : undefined
  const classes = getClassList(el)
  const removed: string[] = []

  classes.forEach((klass) => {
    const matched = !filters.every(f =>
      typeof f === 'string'
        ? f.trim() !== klass.trim()
        : !f.test(klass),
    )
    if (matched)
      removed.push(klass)
  })

  setClass(classes.filter(k => !removed.includes(k)).join(' '))(doc)

  if (cb)
    cb(removed)

  return doc
}
