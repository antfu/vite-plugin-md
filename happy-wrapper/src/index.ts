export { Document, DocumentFragment as Fragment, IElement, IText } from 'happy-dom'

export * from './attributes'
export * from './create'
export * from './diagnostics'
export * from './errors'
export * from './happy-types'
export * from './nodes'
export * from './select'
export * from './type-guards'
export * from './utils'
