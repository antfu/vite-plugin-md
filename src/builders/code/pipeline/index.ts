export * from './addLanguage'
export * from './addClipboard'
export * from './convertBlocksToDomNodes'
export * from './defaultBlocks'
export * from './extractMarkdownItTokens'
export * from './expandCodeBlockVariables'
export * from './inlineStyles'
export * from './highlightLines'
export * from './renderHtml'
export * from './updateLineNumbers'
export * from './useHighlighter'
export * from './userRules'
export * from './updatePreWrapper'
export * from './updateFrontmatterWithCodeBlock'
export * from './updateCodeBlockWrapper'
