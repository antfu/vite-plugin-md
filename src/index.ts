import type { Plugin } from 'vite'
import { createFilter } from '@rollup/pluginutils'
import { createMarkdown } from './markdown'
import { resolveOptions } from './options'
import type { Options } from './types'

function VitePluginMarkdown(userOptions: Options = {}): Plugin {
  const options = resolveOptions(userOptions)
  const markdownToVue = createMarkdown(options)

  const filter = createFilter(
    userOptions.include || /\.md$/,
    userOptions.exclude,
  )

  return {
    name: 'vite-plugin-md',
    enforce: 'pre',
    transform(raw, id) {
      if (!filter(id))
        return
      try {
        return markdownToVue(id, raw)
      }
      catch (e: any) {
        this.error(e)
      }
    },
    async handleHotUpdate(ctx) {
      if (!filter(ctx.file))
        return

      const defaultRead = ctx.read
      ctx.read = async function() {
        return markdownToVue(ctx.file, await defaultRead())
      }
    },
  }
}

export default VitePluginMarkdown
