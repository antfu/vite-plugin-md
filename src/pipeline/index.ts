export * from './applyMarkdownItOptions'
export * from './baseStyling'
export * from './composeSfcBlocks'
export * from './createParser'
export * from './convertHandlerToTask'
export * from './convertToDom'
export * from './escapeCodeTagInterpolation'
export * from './extractBlocks'
export * from './extractFrontmatter'
export * from './finalize'
export * from './pipelineUtilityFunctions'
export * from './frontmatterPreprocess'
export * from './loadMarkdownItPlugins'
export * from './parse'
export * from './repairFrontmatterLinks'
export * from './transformsBefore'
export * from './wrapHtml'
