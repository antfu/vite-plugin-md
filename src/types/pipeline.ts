import type MarkdownIt from 'markdown-it'
import type { MaybeRef } from '@vueuse/core'
import type * as TE from 'fp-ts/TaskEither'
import type { UserConfig } from 'vite'
import type { Either } from 'fp-ts/lib/Either'
import type { Fragment, IElement } from 'happy-wrapper'
import type { EnumValues, Frontmatter, MetaProperty, ResolvedOptions } from './core'

export enum PipelineStage {
  /**
   * Initialized with incoming filename, config, options,
   * available events (core and provided by builders).
   */
  initialize = 'initialize',

  /**
   * All frontmatter has been extracted from default values and page values
   * but no mapping has been done yet.
   *
   * Note: this is the event hook which the included `meta` builder connects
   * to and it in turn _provides_ a `metaMapped` hook.
   */
  metaExtracted = 'metaExtracted',

  /**
   * The **MarkdownIt** parser is initialized, all builders
   * connecting at this point will receive a valid `md` parser
   * object so that they can participate in MD-to-HTML parsing.
   */
  parser = 'parser',
  /**
   * The **MarkdownIt** parser is initialized and all builders
   * have been able to apply their customizations to it.
   */
  parsed = 'parsed',

  /**
   * The HTML has been converted to a HappyDom tree to allow
   * interested parties to manipulate contents using DOM based
   * queries.
   */
  dom = 'dom',

  /**
   * SFC blocks (template, script, and an array of customBlocks) are ready for
   * builders to inspect/mutate/etc.
   */
  sfcBlocksExtracted = 'sfcBlocksExtracted',

  /**
   * All mutations of page are complete; builders can hook into this stage but
   * will _not_ be able to mutate at this stage.
   */
  closeout = 'closeout',
}

export type IPipelineStage = EnumValues<PipelineStage>

export interface RulesUse {
  ruleName: string
  usage: 'adds' | 'patches' | 'modifies'
  description?: string
}

export type PipelineInitializer = (i?: Pipeline<PipelineStage.initialize>) => Pipeline<PipelineStage.initialize>

export interface BuilderRegistration<O extends BuilderOptions, S extends IPipelineStage> {
  name: string
  description?: string
  /** The lifecycle event/hook which this builder will respond to */
  lifecycle: S
  /**
   * The builder's handler function which receives the _payload_ for the
   * event lifecycle hook configured and then is allowed to mutate these
   * properties and pass back a similarly structured object to continue
   * on in that pipeline stage.
   */
  handler: BuilderHandler<O, S>

  /**
   * The options _specific_ to the builder
   */
  options: O

  /**
   * This isn't strictly required, but it is nice to express which rules you have used
   * modified, or added from the MarkdownIt parser.
   *
   * Note: builders should try to avoid mutating core rules; if they need a modification
   * for their purposes consider _monkey patching_ the rule so that downstream rules
   * have a better understanding of current rule state.
   */
  parserRules?: RulesUse[]

  /**
   * If this plugin needs to modify the configuration in some way at initialization
   * it can add a function here to do that. In most cases, the builder can simply
   * wait for their event hook to be called (at which point they will get the configuration
   * passed to them).
   */
  initializer?: BuilderHandler<O, PipelineStage.initialize>
}

export type Parser<S extends IPipelineStage> = S extends 'parser' | 'parsed' | 'dom' | 'sfcBlocksExtracted' | 'closeout'
  ? {
      /** the **MarkdownIT** parser instance */
      parser: MarkdownIt

    }
  : {}

/**
 * The **Route** custom-block can be configured with these properties.
 *
 * Note: this is best done with the _meta_ builder
 */
export interface RouteConfig {
  name?: string
  path?: string
  meta?: Record<string, any>
}

export interface LinkProperty {
  rel?: string
  href?: string
  integrity?: string
  crossorigin?: string
  [key: string]: unknown
}

export interface StyleProperty {
  type?: string
  [key: string]: unknown
}

/**
 * The <base> element specifies the base URL and/or target for all
 * relative URLs in a page.
 */
export interface BaseProperty {
  href?: string
  target?: string
}

export interface ScriptProperty {
  /**
   * For classic scripts, if the async attribute is present, then the classic
   * script will be fetched in parallel to parsing and evaluated as soon as it is
   * available.
   *
   * For module scripts, if the async attribute is present then the scripts and
   * all their dependencies will be executed in the defer queue, therefore they will
   * get fetched in parallel to parsing and evaluated as soon as they are available.
   */
  async?: boolean
  crossorigin?: string
  /** only to be used alongside the `src` attribute */
  defer?: boolean
  /** Provides a hint of the relative priority to use when fetching an external script.  */
  fetchPriority?: 'high' | 'low' | 'auto'

  integrity?: string
  /**
   * This Boolean attribute is set to indicate that the script should not be executed
   * in browsers that support ES2015 modules â€” in effect, this can be used to serve
   * fallback scripts to older browsers that do not support modular JavaScript code.
   */
  nomodule?: boolean
  nonce?: string
  referencePolicy?: 'no-referrer' | 'no-referrer-when-downgrade' | 'origin' | 'origin-when-cross-origin' | 'same-origin' | 'strict-origin' | 'strict-origin-when-cross-origin' | 'unsafe-url'
  src?: string
  type?: string
  [key: string]: unknown
}

/**
 * The properties destined for the HEAD block in the HTML
 */
export interface HeadProps {
  title?: MaybeRef<string>
  meta?: MaybeRef<MetaProperty[]>
  link?: MaybeRef<LinkProperty[]>
  base?: MaybeRef<BaseProperty[]>
  style?: MaybeRef<StyleProperty[]>
  script?: MaybeRef<ScriptProperty[]>
  htmlAttrs?: MaybeRef<Record<string, unknown>[]>
  bodyAttrs?: MaybeRef<Record<string, unknown>[]>
  [key: string]: unknown
}

export interface PipelineUtilityFunctions {
  /**
   * Adds a `<link>` to the page's header section
   */
  addLink: (link: LinkProperty) => void
  /**
   * Adds a `<script>` reference to the page's header section
   */
  addScriptReference: (script: ScriptProperty) => void

  /**
   * Allows the addition of code which will be brought into the
   * `<script setup>` block if using VueJS 3.x and into a normal
   * `<script>` block in VueJS2
   */
  addCodeBlock: (name: string, script: string, forVue2?: string[] | undefined) => void

  /**
   * Adds a `<style>` reference to the page's header section
   */
  addStyleReference: (style: StyleProperty) => void
  /**
   * Adds a VueJS `<script>` block to the HTML (which VueJS will eventually place in HEAD). A style block should be named so that downstream consumers
   * can -- potentially -- override or further modify the style.
   */
  addStyleBlock: (name: string, style: IElement | string) => void
}

export type MetaExtracted<S extends IPipelineStage> = S extends 'initialize'
  ? {}
  : {
      /** the frontmatter metadata */
      frontmatter: Frontmatter

      /**
       * The markdown content (after extracting frontmatter)
       */
      md: string

      /**
       * Meta properties that will be put into the HEAD section
       *
       */
      // TODO: this should be removed as it's duplicative with what's in `head`
      meta: MetaProperty[]

      excerpt?: string

    } & PipelineUtilityFunctions

export type HtmlContent<S extends IPipelineStage> = S extends 'parsed' | 'sfcBlocksExtracted' | 'closeout'
  ? {
      /**
       * the HTML produced from MD content (and using parser rules passed in)
       */
      html: string
    }
  : S extends 'dom'
    ? {
        /**
         * the HTML wrapped into a HappyDom fragment
         */
        html: Fragment

        /**
         * If any code blocks were found on the page then their languages will be represented
         * here as a `Set<string>`
         */
        fencedLanguages: Set<string>
      }
    : {}

export type Blocks<S extends IPipelineStage> = S extends 'sfcBlocksExtracted' | 'closeout'
  ? {
      /**
       * all hoisted scripts
       */
      hoistedScripts: string[]

      /** the SFC's template block (aka, html content) */
      templateBlock: string

      /** the `<script [setup] ...>` block */
      scriptBlock: string

      /** any other top-level SFC blocks besides "template" and "script" */
      customBlocks: string[]
    }
  : {}

export type Completed<S extends IPipelineStage> = S extends 'closeout'
  ? {
      /** the finalized component in string form */
      component: string
    }
  : {}

export type Pipeline<S extends IPipelineStage> = {
  fileName: string
  /** the raw content in the file being processed */
  content: string
  /** the `vite-plugin-md` options */
  options: ResolvedOptions
  /** the Vite config */
  viteConfig: UserConfig

  /**
   * All properties which are destined for the HEAD section of the HTML
   */
  head: HeadProps
  /**
   * Meta properties associated with a page's route; typically used
   * and managed with the "meta" builder.
   */
  routeMeta?: RouteConfig

  /**
   * Indicates which _languages_ were found on the page; this property typically
   * shouldn't be set but rather is managed by the `code()` builder and provided
   * for contextual decisions.
   */
  codeBlockLanguages: {
    /** the language setting the page author used */
    langsRequested: string[]
    /** the language used to parse with the highlighter */
    langsUsed: string[]
  }

  /**
   * A store for _named_ VueJS `<style>` blocks which will be injected
   * into the SFC component at the appropriate time.
   */
  vueStyleBlocks: Record<string, IElement>

  /**
   * Provides a _named_ set of code blocks which will be injected into
   * the VueJS's SFC's `<script setup>` section for Vue 3 and in a
   * `<script>` block. All blocks will be setup for Typescript.
   *
   * Note: contributors may optionally include additional trailing lines
   * for Vue2; this will allows you to export variables you've defined.
   */
  vueCodeBlocks: Record<string, string | [base: string, vue2Exports: string[]]>
} & Parser<S> & MetaExtracted<S> & HtmlContent<S> & Blocks<S> & Completed<S>

/**
 * The Builder's event listener/handler
 */
export type BuilderHandler<
  O extends BuilderOptions,
  S extends IPipelineStage,
> = (payload: Pipeline<S>, options: O) => Promise<Pipeline<S>>

/**
 * Users configure a `BuilderHandler` and we wrap this up functionality
 * with a higher-order `BuilderTask`.
 *
 * @returns TaskEither<string, Pipeline<S>>
 */
export type BuilderTask<
  S extends IPipelineStage,
> = () => (payload: Pipeline<S>) => TE.TaskEither<string, Pipeline<S>>

/**
 * Builder's must provide an export which meets this API constraint. Basic
 * structure of this higher order function is:
 *
 * - options( ) -> register( ) -> { handler( payload ) -> payload }
 */
export type BuilderApi<O extends {}, S extends IPipelineStage> = (options?: O) => () => BuilderRegistration<O, S>

export type InlineBuilder = <N extends string, L extends IPipelineStage>(name: N, lifecycle: L) => (payload: Pipeline<L>) => Pipeline<L>

/**
 * Builder options are expected to be a key/value dictionary but must
 * be allowed to be an empty object
 * */
export type BuilderOptions = Record<string, any> | {}

export type BuilderConfig = Record<IPipelineStage, BuilderRegistration<any, any>[] | []>

/**
 * Carries an Either<T> condition which is either:
 * - a _string_ error condition
 * - a `Pipeline<S>` value
 */
export type PipeEither<S extends IPipelineStage> = Either<string, Pipeline<S>>

/**
 * Carries an `TaskEither<T>` condition which is either:
 * - a _string_ error condition
 * - a `() => Promise<Pipeline<S>>` task
 */
export type PipeTask<S extends IPipelineStage> = TE.TaskEither<string, Pipeline<S>>

/**
 * A pipeline payload or either an async `PipeTask<S>` or a synchronous `PipeEither<S>`
 */
export type PipelinePayload<S extends IPipelineStage> = PipeTask<S> | PipeEither<S>

/**
 * A _synchronous_ transformer function which:
 *
 * - receives a payload of `PipeEither<F>`, and
 * - converts it to a type of `PipeEither<T>`
 */
export type SyncPipelineTransformer<
  F extends IPipelineStage,
  T extends IPipelineStage,
> = (payload: PipeTask<F>) => PipeTask<T>

/**
* An _asynchronous_ transformer function which:
*
* - receives a payload of `PipeTask<F>` (async) or `PipeEither<F>` (sync)
* - converts it to a type of `PipeTask<T>`
*/
export type AsyncPipelineTransformer<
  F extends IPipelineStage,
  T extends IPipelineStage,
> = (payload: PipeTask<F>) => PipeTask<T>
