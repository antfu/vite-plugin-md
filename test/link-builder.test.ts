import { readFile } from 'fs/promises'
import { beforeAll, describe, expect, it } from 'vitest'
import { Window } from 'happy-dom'
import { composeSfcBlocks } from '../src/pipeline'
import { link } from '../src/index'

const window = new Window()
const document = window.document

let md = ''

describe('link testing', () => {
  beforeAll(async () => {
    md = await readFile('test/fixtures/links.md', 'utf-8')
  })

  it('component snapshot is consistent', async () => {
    const sfc = await composeSfcBlocks('links.md', md, { builders: [link()] })
    expect(sfc.component).toMatchSnapshot()
  })

  it('class links are found in HTML text', async () => {
    const { html } = await composeSfcBlocks('links.md', md, { builders: [link()] })
    expect(html.includes('internal-link')).toBeTruthy()
    expect(html.includes('external-link')).toBeTruthy()
  })

  it('internal and external classes are found in DOM', async () => {
    const sfc = await composeSfcBlocks('links.md', md, { builders: [link()] })

    document.body.innerHTML = sfc.html
    const internalLinks = document.querySelectorAll('.internal-link')
    const externalLinks = document.querySelectorAll('.external-link')

    for (const l of internalLinks)
      expect(l.classList.value, `class list should have had "internal-link": ${l.classList.value}`).toContain('internal-link')

    expect(internalLinks.map(i => i.textContent)).toContain('internal link')
    expect(internalLinks.map(i => i.textContent)).not.toContain('external link')

    for (const l of externalLinks)
      expect(l.classList.value, `class list should have had "external-link": ${l.classList.value}`).toContain('external-link')

    expect(externalLinks.map(i => i.textContent)).toContain('external link')
    expect(externalLinks.map(i => i.textContent)).not.toContain('internal link')
  })

  it('content-type classes are brought in appropriately', async () => {
    const sfc = await composeSfcBlocks('', md, { builders: [link()] })
    document.body.innerHTML = sfc.html
    const imageLinks = document.querySelectorAll('.image-reference')
    const docLinks = document.querySelectorAll('.doc-reference')
    const mailTo = document.querySelectorAll('.mailto-link')

    expect(imageLinks.map(i => i.textContent)).toContain('images')
    expect(docLinks.map(i => i.textContent)).toContain('documents')
    expect(mailTo.map(i => i.textContent)).toContain('contact us')
  })

  it('custom rules add classes as expected', async () => {
    const sfc = await composeSfcBlocks('', md, {
      builders: [link({
        ruleBasedClasses: [[/colors\.com/, 'colorful']],
      })],
    })

    document.body.innerHTML = sfc.html

    const colorful = document.querySelectorAll('.colorful')

    expect(colorful.length).toBe(3)
    expect(colorful.map(i => i.textContent)).toContain('red')
    expect(colorful.map(i => i.textContent)).toContain('blue')
    expect(colorful.map(i => i.textContent)).toContain('green')
  })

  it('internal index route reference to a markdown file is cleaned up', async () => {
    const sfc = await composeSfcBlocks('', md, { builders: [link({ useRouterLinks: false })] })
    document.body.innerHTML = sfc.html
    const internal = document.querySelectorAll('.internal-link')
    const indexRoute = internal.find(i => i.textContent === 'index route')
    expect(indexRoute).not.toBeUndefined()
    expect(indexRoute?.getAttribute('href')).toEqual('/foobar/')

    // double check that conversion to router link works in the same way
    const sfc2 = await composeSfcBlocks('', md, { builders: [link()] })
    document.body.innerHTML = sfc2.html
    const internal2 = document.querySelectorAll('.internal-link')
    const indexRoute2 = internal2.find(i => i.textContent === 'index route')
    expect(indexRoute2).not.toBeUndefined()
    expect(indexRoute2?.getAttribute('href')).toBeNull()
    expect(indexRoute2?.getAttribute('to')).toEqual('/foobar/')
  })

  it('internal non-index route with MD in href is shortened to route path', async () => {
    const sfc = await composeSfcBlocks('repo/src/pages/current.md', md, { builders: [link({ useRouterLinks: false })] })
    document.body.innerHTML = sfc.html
    const internal = document.querySelectorAll('.internal-link')
    const nonIndexRoute = internal.find(i => i.textContent === 'non-index route')
    expect(nonIndexRoute).not.toBeUndefined()
    expect(nonIndexRoute?.getAttribute('href')).toEqual('foo/bar')
  })

  it('external routes with .md reference are left as is', async () => {
    const sfc = await composeSfcBlocks('', md, { builders: [link()] })
    document.body.innerHTML = sfc.html
    const external = document.querySelectorAll('.external-link')
    const indexRoute = external.find(i => i.textContent === 'external index routes')
    expect(indexRoute).not.toBeUndefined()
    expect(indexRoute?.getAttribute('href')).toEqual('https://dev.null/foo/index.md')

    const nonIndexRoute = external.find(i => i.textContent === 'external non-index routes')
    expect(nonIndexRoute).not.toBeUndefined()
    expect(nonIndexRoute?.getAttribute('href')).toEqual('https://dev.null/foo/bar.md')
  })

  it('"base" option set changes link resolution for relative links', async () => {
    const sfc = await composeSfcBlocks('', md, {
      builders: [
        link({ relativeLinkClass: 'relative-link' }),
      ],
    }, { base: 'one' })
    document.body.innerHTML = sfc.html

    const links = document.querySelectorAll('.relative-link')
    expect(links.length).toBeGreaterThan(0)

    for (const l of links) {
      expect(
        l.getAttribute('to').startsWith('/one'),
        `when we have a "base" of "one" all internal URLs should start with that URL but the URL was "${l.getAttribute('to')}"\n`,
      ).toBeTruthy()
    }
  })

  it('"base" option set changes link resolution for fully qualified local links', async () => {
    const sfc = await composeSfcBlocks('', md, { builders: [link({ fullyQualifiedLinkClass: 'fq-link' })] }, { base: 'one' })
    document.body.innerHTML = sfc.html

    const links = document.querySelectorAll('.fq-link')
    expect(links.length).toBeGreaterThan(0)

    for (const l of links) {
      expect(
        l.getAttribute('to').startsWith('/one'),
        `when we have a "base" of "foo" all internal URLs should start with that URL. Url was:  ${l.getAttribute('to')}\n`,
      ).toBeTruthy()
    }
  })

  it('internal routes are converted to `<router-link>` elements by default', async () => {
    const sfc = await composeSfcBlocks('test/fixtures/links.md', md, { builders: [link()] })
    document.body.innerHTML = sfc.html

    const internal = document.querySelectorAll('.internal-link')
    const router = document.querySelectorAll('.router-link')

    expect(internal.length).toEqual(router.length)

    router.forEach(t =>
      expect(t.tagName, 'the tagname should have been converted from <a> to <router-link>').toEqual('ROUTER-LINK'),
    )
    router.forEach(t =>
      expect(t.getAttribute('href'), 'the "href" should be empty in favor of a "to" prop').toBeNull(),
    )
    router.forEach(t =>
      expect(t.getAttribute('to'), 'the "to" property should be set for <router-link>').toBeTypeOf('string'))
  })

  it('the "_base" attribute is removed prior to rendering HTML of router links', async () => {
    const sfc = await composeSfcBlocks(
      'test/fixtures/links.md',
      md,
      {
        builders: [link()],
      },
    )
    document.body.innerHTML = sfc.html

    const router = document.querySelectorAll('.router-link')

    router.forEach(t =>
      expect(t.getAttribute('_base'), 'the _base attribute is set during the build process to ensure a relative path but should be removed in final rendering').toBeNull(),
    )
  })

  it('router links which are relative are resolved to a valid route', async () => {
    const sfc = await composeSfcBlocks(
      'test/fixtures/links.md',
      md,
      {
        builders: [link()],
      },
    )
    document.body.innerHTML = sfc.html

    const router = document.querySelectorAll('.router-link')

    router.forEach(t => expect(t.getAttribute('to').startsWith('test/fixtures')))
  })

  it('router links add in "base" when present', async () => {
    const sfc = await composeSfcBlocks(
      'test/fixtures/links.md',
      md,
      {
        builders: [link()],
      },
      { base: 'foobar' },
    )
    document.body.innerHTML = sfc.html

    const router = document.querySelectorAll('.router-link')

    router.forEach(t => expect(t.getAttribute('to').startsWith('foobar/test/fixtures')))
  })
})
