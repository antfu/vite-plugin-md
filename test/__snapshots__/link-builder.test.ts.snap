// Vitest Snapshot v1

exports[`link testing > component snapshot is consistent 1`] = `
"<route lang=\\"json\\">{}</route>

<template>
  <div class=\\"markdown-body\\"><h1>Links</h1>
<ul>
<li>An <a href=\\"https://google.com\\" class=\\" external-link\\" target=\\"_blank\\" rel=\\"noreferrer noopenner\\">external link</a> is one which links to another site whereas an <router-link class=\\" internal-link router-link\\" to=\\"/simple\\">internal link</router-link> links to content on the same site.</li>
<li>Within the category of “internal links” we have <router-link class=\\" internal-link router-link\\" to=\\"/foo/bar\\">fully qualified links</router-link> and <router-link class=\\" internal-link router-link\\" to=\\"/foo/bar\\">relative links</router-link> where with relative links we must ensure that the current page’s route is understood and incorporated into the link</li>
<li>Check out <router-link class=\\" internal-link anchor-tag router-link\\" to=\\"/#section-2\\">Section 2</router-link> for the good stuff … well actually to be fair it’s just an example of an “anchor tag” where we’re linking to the another part of the same page</li>
<li>External links are great but some aren’t safe: <a href=\\"http://bad-juju.com\\" class=\\" external-link insecure\\" target=\\"_blank\\" rel=\\"noreferrer noopenner\\">scary link</a></li>
</ul>
<h2>Markdown file in a link</h2>
<ul>
<li>When we reference an index MD file the <router-link class=\\" internal-link router-link\\" to=\\"/foobar/\\">index route</router-link> is left with just the path component, no file component</li>
<li>Similarly if an <em>internal</em> link to a <router-link class=\\" internal-link router-link\\" to=\\"/foo/bar\\">non-index route</router-link> has a <code class=\\"\\">.md</code> reference in it … this is cleaned up.</li>
<li>NOTE: <em>external</em> links for both <a href=\\"https://dev.null/foo/index.md\\" class=\\" external-link\\" target=\\"_blank\\" rel=\\"noreferrer noopenner\\">external index routes</a> and <a href=\\"https://dev.null/foo/bar.md\\" class=\\" external-link\\" target=\\"_blank\\" rel=\\"noreferrer noopenner\\">external non-index routes</a> are not changed (but shouldn’t really be happening a lot)</li>
</ul>
<h2>Content Types</h2>
<ul>
<li>If we link to <a href=\\"https://my-photos.com/profile.jpg\\" class=\\" external-link image-reference\\" target=\\"_blank\\" rel=\\"noreferrer noopenner\\">images</a>, <a href=\\"https://xyz.com/profile.doc\\" class=\\" external-link doc-reference\\" target=\\"_blank\\" rel=\\"noreferrer noopenner\\">documents</a>, or <a href=\\"https://example.com/foobar.wasm\\" class=\\" external-link\\" target=\\"_blank\\" rel=\\"noreferrer noopenner\\">code</a> we should get classes which indicate the content type</li>
<li>My favorite colors are <a href=\\"https://colors.com/red\\" class=\\" external-link\\" target=\\"_blank\\" rel=\\"noreferrer noopenner\\">red</a>, <a href=\\"https://colors.com/blue\\" class=\\" external-link\\" target=\\"_blank\\" rel=\\"noreferrer noopenner\\">blue</a>, and <a href=\\"https://colors.com/green\\" class=\\" external-link\\" target=\\"_blank\\" rel=\\"noreferrer noopenner\\">green</a></li>
</ul>
<h2>Router Awareness</h2>
<ul>
<li>we can convert all internal links from normal <code class=\\"\\">&lt;a href=&quot;xyz&quot;&gt;</code> tags to <code class=\\"\\">&lt;router-link to=&quot;xyz&quot;&gt;</code> tags instead; this feature is on by default</li>
<li>this is helpful for a PWA as you get the benefits of</li>
</ul>
<h2>Section 2</h2>
<p>The good stuff</p>
<p><a href=\\"mailto:yur-da-best@dev.null\\" class=\\" mailto-link\\">contact us</a></p>
</div>
</template>

<script setup>
  defineExpose({ frontmatter: {\\"title\\":\\"Links Testing\\"} })
  
 const title = \\"Links Testing\\"
 const excerpt = undefined
</script>

<script>
  /** frontmatter meta-data for MD page **/
  export const frontmatter = {\\"title\\":\\"Links Testing\\"}
</script>
"
`;
